src/source/mod.rs

```diff
@@ -20,40 +21,112 @@ use rustc_hash::FxHashMap;
 use unic_langid::LanguageIdentifier;
 
 pub type RcResource = Rc<FluentResource>;
-pub type ResourceOption = Option<RcResource>;
-pub type ResourceFuture = Shared<Pin<Box<dyn Future<Output = ResourceOption>>>>;
+
+/// An option type whose None variant is either optional or required.
+///
+/// This behaves similarly to the standard-library [`Option`] type
+/// except that there are two [`None`]-like variants:
+/// [`ResourceOption::MissingOptional`] and [`ResourceOption::MissingRequired`].
+#[derive(Clone, Debug)]
+pub enum ResourceOption<T> {
```

Comment:
`question`: Why did you introduce `ResourceOption` as a generic here? What else can it be generalized over?

Reply:
When I added `as_ref()` in particular I needed the generic parameter so that a `ResourceOption` could hold a reference instead of an owned type all the time. 

I added these because I thought I needed them in Gecko, but it's looking like I will not need them. I should only need the `to_option()` function, which I will use the `From` trait for.

Reply:
I initially made it generic when I added as_ref() function, so that the ResourceOption could be generic over &T and not just T.

I did this initially to get something compiling in Gecko that was calling as_ref() on what used to be just a regular Option type but is now a ResourceOption.

But this problem is solved by having a to_option() function, since that particular context in Gecko does no care if the resource is required or not. I believe I can remove the as_ref() and map() functions and keep just the to_option() in the form of the From trait like you suggested.